src/main/scala/is/hail/io/plink/ExportBedBimFam.scala

```diff
@@ -1,34 +1,71 @@
 package is.hail.io.plink
 
+import is.hail.sparkextras._
+import is.hail.expr._
+import is.hail.annotations._
 import is.hail.variant._
 
 object ExportBedBimFam {
 
   val gtMap = Array(1, 3, 2, 0)
 
-  def makeBedRow(gs: Iterable[Genotype], n: Int): Array[Byte] = {
-    val gts = gs.hardCallIterator
-
-    val nBytes = (n + 3) / 4
-    val a = Array.ofDim[Byte](nBytes)
-    var b = 0
-    var k = 0
-    while (k < n) {
-      b |= gtMap(gts.next() + 1) << ((k & 3) * 2)
-      if ((k & 3) == 3) {
-        a(k >> 2) = b.toByte
-        b = 0
+  def bedRowTransformer(nSamples: Int, rowType: TStruct): Iterator[RegionValue] => Iterator[Array[Byte]] = { it =>
+    val hcv = HardCallView(rowType)
+    val rvb = new RegionValueBuilder()
+    val rv2 = RegionValue()
+    val gtMap = Array(3, 2, 0)
+
+    it.map { rv =>
+      hcv.setRegion(rv)
+      rvb.set(rv.region)
+
+      val nBytes = (nSamples + 3) / 4
+      val a = new Array[Byte](nBytes)
+      var b = 0
+      var k = 0
+      while (k < nSamples) {
+        hcv.setGenotype(k)
+        val gt = if (hcv.hasGT) gtMap(hcv.getGT) else 1
+        b |= gt << ((k & 3) * 2)
+        if ((k & 3) == 3) {
+          a(k >> 2) = b.toByte
+          b = 0
+        }
+        k += 1
       }
-      k += 1
-    }
-    if ((k & 3) > 0)
-      a(nBytes - 1) = b.toByte
+      if ((k & 3) > 0)
+        a(nBytes - 1) = b.toByte
 
-    a
+      // FIXME: NO BYTE ARRAYS, go directly through writePartitions
+      a
+    }
   }
 
-  def makeBimRow(v: Variant): String = {
-    val id = s"${v.contig}:${v.start}:${v.ref}:${v.alt}"
-    s"""${v.contig}\t$id\t0\t${v.start}\t${v.alt}\t${v.ref}"""
+  def bimRowTransformer(rowType: TStruct): Iterator[RegionValue] => Iterator[String] = { it =>
+    val vIdx = rowType.fieldIdx("v")
```

Comment:
We should break this out as a VariantView.  If you want to do that as a separate change let me know and I'll  read through this a bit more closely.